### ISSUE TITLE
Checking a permanently denied permission returns false unless requesting the permission is done first, is this a bug or what?

### ISSUE THREAD

**Issue by Haidar0096** [Baseflow/flutter-permission-handler#647]
There are 2 methods in the library to check for a permission's status, for example in case of the microphone we can call:

- `PermissionStatus status1 = await Permission.microphone.status //call 1`

- `PermissionStatus status2 = await Permission.microphone.request() //call 2`

suppose a user presses the 'dont ask again' when he is prompted to grant the permission.
 Now  If:
 - we make **call 1** to get the permission status, we should expect to get `PermissionStatus.permanentlyDenied` but the result is `PermissionStatus.denied`
 - we make **call 2** to get the permission status, we expect to get `PermissionStatus.permanentlyDenied` and without the dialouge appearing again to the user, and this is exactly what happens
 

So **call 2** works as intended but **call 1** works as if there are only 2 states for a permission that can be returned from this call: granted or denied. But then how can we get the real status of the permission without calling `request()` in this case? Is this a bug or is this the intended behaviour? Sometimes I need to just check the status of the permission but I don't want to open a permission dialouge to the user, how to do that?

**Comment by mvanbeusekom** [comment]
This is working as intended. The reason is explained in this [issue](https://github.com/Baseflow/flutter-permission-handler/issues/568#issuecomment-849494650) and also on our [Wiki](https://github.com/Baseflow/flutter-permission-handler/wiki/Changes-in-6.0.0#breaking-changes).

TLDR: Android doesn't support checking the difference between "denied" permissions and "permanently denied" permissions.

**Comment by Haidar0096** [comment]
@mvanbeusekom thanks that thread explained things alot, It would still be better if we could know the real detailed status on android, it does not make sense to call `request()` even If I want to check the permission, sometimes the permission will be just deniedOnce and then a dialouge will be show to the user although I don't want that, is there a workaround on this problem?

**Comment by mvanbeusekom** [comment]
I agree that would be better but unfortunately Google didn't make this available (they don't explain why). So we will simply have to check permissions and if not equal to "granted" we should request them.

Pre Android API 30 there was a workaround but from API 30 and higher this workaround causes more problems then it solves so we decided to remove it.

**Comment by Haidar0096** [comment]
@mvanbeusekom I made this diagram that models **a way** to check if the permission is permanently denied

- Is this what you meant in the other thread you linked above when you talked about using `shouldShowRequestPermissionRationale`?

- Is this a correct approach in your opinion?

![image](https://user-images.githubusercontent.com/49842031/129263241-f2ab02a7-715b-49ce-ae4f-ed726e355956.png)

**Comment by mvanbeusekom** [comment]
@Haidar0096 pre-Android API 30 this is indeed correct. However since Android API 30 the new "Only once" permission has been added which resets the permissions as soon as your app closes. This makes the `checkPermission` and `shouldShowPermissionRationale` behave exactly the same way as if the first time the App starts. 

According to your diagram this will always result in a permission denied forever state until the user manually changes permissions in the app settings.

**Comment by Haidar0096** [comment]
@mvanbeusekom Ah I get you. The problem is I am using this functionality with the BLoC pattern in my app and it is causing real problems of the BLoC's  state, I guess I should abandon the idea of checking the **real** status of the permission without requesting it, and the only way to get rid of the headache is to not allow the user to enter the app at all unless he gives the permission at start. Thanks and sorry for the extensive questions from my side.

**Comment by mvanbeusekom** [comment]
Hi @Haidar0096,

The way to handle this would be to first check the permission status and when it returns "denied", request permissions.

When the user denied permissions forever Android (and Apple) will not show a permission dialog and the `request()` method will return the correct `PermissionStatus.permanentlyDenied` status.

Something like this should work:

```dart
￼PermissionStatus getPermission() {
  var status = Permission.microphone.status;

  if (status == PermissionStatus.denied) {
    status = Permission.microphone.request();
  }

  return status;
}
```
